src/tests/JIT/Regression/JitBlue/Runtime_39581/Runtime_39581.il

```diff
@@ -0,0 +1,174 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+//
+// The test exposes the issue when:
+//
+//   1) methods Callee and Caller have the following structure:
+//
+//      S Callee() { produce and return an instance of S }
+//
+//      S Caller() { return Callee(); }
+//
+//   2) S is a value type with non-GC fields and has such size that it
+//      must passed via return buffer
+//
+//   3) call to Callee is transformed to tail call via helper:
+//
+//      S result;
+//      DispatchTailCalls(&IL_STUB_CallTailCallTarget, (IntPtr)&result, _AddressOfReturnAddress());
+//      return result;
+//
+//   4) Caller is invoked via Reflection.
+//
+// During morph phase the JIT discovers that both Caller and Callee have the return buffer arguments and
+// falsely assumes that the value can be directly passed from Caller to Callee.
+// Here is a problem: DispatchTailCalls helper expects the RetValue argument (in this case, return buffer argument)
+// to always point on the stack. However, during a reflection call the return buffer for S value type can be placed on the GC heap (see condition 2 above).
```

Comment:
// to always point to the stack. However, during a reflection call the return buffer for S value type can be placed on the GC heap (see condition 2 above).